### ISSUE TITLE
python310 is broken on x86_64-darwin

### ISSUE THREAD

**Issue by veprbl** [NixOS/nixpkgs#154120]
See https://hydra.nixos.org/build/163469637
`python310Packages` are now a thing: #152307

**Comment by prusnak** [comment]
python310 seems to build on aarch64-darwin just fine

**Comment by mweinelt** [comment]
cc @NixOS/darwin-maintainers

**Comment by smaret** [comment]
This seems related: https://discuss.python.org/t/cpython-compilation-get-stuck-at-generate-posix-vars/4987

**Comment by starcraft66** [comment]
Could this be related to codesigning somehow?
The log shows:
```
/nix/store/b0q4wjdw6xwhfsqy8ab106cmhn5xrr8p-bash-5.1-p12/bin/bash: line 6: 93471 Killed: 9               DYLD_LIBRARY_PATH=/private/tmp/nix-build-python3-3.10.1.drv-0/Python-3.10.1 ./python.exe -E -S -m sysconfig --generate-posix-vars
```

Often times when something just gets killed like that on darwin it has to do with the binary not being signed.

**Comment by toonn** [comment]
I don't believe code signing is a thing on x86_64-darwin, as far as Nixpkgs is concerned? Since the versions involved are 10.12 and 10.13.

**Comment by veprbl** [comment]
Running the python.exe manually (even without the arguments), I see `shared_region: enter(0x64ec414cc8604bf,0x64ec414cdf0b637,0x0,7,0,1): vm_map_enter(0x7fff00000000,0xffe00000,0x64ec414bea8c89f) error 0x3` in the kernel log and no crash dump. The google pointed me to https://github.com/golang/go/issues/895, so I've tried to use dtrace to see what is being done.

```
dtrace -n 'fbt::vm_shared_region_enter:entry{trace(pid);trace(execname);printf("%xd %xd %xd\n", args[0]->hdr.links.start, args[0]->hdr.links.end, args[0]->size);stack();}' -n 'fbt::vm_shared_region_enter:return{trace(arg0);trace(arg1);}'
```
returns a few entries and only one is returning 0x3
<detail>
```
dtrace: description 'fbt::vm_shared_region_enter:entry' matched 1 probe
dtrace: description 'fbt::vm_shared_region_enter:return' matched 1 probe
CPU     ID                    FUNCTION:NAME
  0 278470     vm_shared_region_enter:entry      6943  zsh                              107692000d 7fffffe00000d 10d0000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  0 278471    vm_shared_region_enter:return              3642                3
  7 278470     vm_shared_region_enter:entry      6944  zsh                              4000000d 7fffffe00000d 4ba000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  7 278471    vm_shared_region_enter:return              3642                0
  0 278470     vm_shared_region_enter:entry      6946  zsh                              107f20000d 7fffffe00000d d8000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  0 278471    vm_shared_region_enter:return              3642                0
  0 278470     vm_shared_region_enter:entry      6949  zsh                              10fb3b000d 7fffffe00000d 121000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  0 278471    vm_shared_region_enter:return              3642                0
  0 278470     vm_shared_region_enter:entry      6951  zsh                              105b6a000d 7fffffe00000d 45d000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  0 278471    vm_shared_region_enter:return              3642                0
  2 278470     vm_shared_region_enter:entry      6952  zsh                              10910e000d 7fffffe00000d 45d000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16

  2 278471    vm_shared_region_enter:return              3642                0
  4 278470     vm_shared_region_enter:entry      6953  zsh                              105749000d 7fffffe00000d 45d000d

              kernel`vm_map_exec+0x17f
              kernel`0xffffff8019eaf1c0+0x10a7
              kernel`0xffffff8019eb61c0+0x642
              kernel`__mac_execve+0x227
              kernel`execve+0x59
              kernel`unix_syscall64+0x27d
              kernel`hndl_unix_scall64+0x16
```
</detail>

So, I guess our python executable doesn't even start, so it's not the go bug. Then it's something to do with loading of the mach-o files themselves?

**Comment by K900** [comment]
Presumably by sheer luck today's 12.2 beta added a better log message here:

```
python3> dyld[20140]: dyld cache '/System/Library/dyld/dyld_shared_cache_x86_64h' not loaded: syscall to map cache into shared region failed
python3> dyld[20140]: Library not loaded: /usr/lib/libSystem.B.dylib
python3>   Referenced from: /private/tmp/nix-build-python3-3.10.1.drv-0/Python-3.10.1/python.exe
python3>   Reason: tried: '/private/tmp/nix-build-python3-3.10.1.drv-0/Python-3.10.1/libSystem.B.dylib' (no such file), '/usr/lib/libSystem.B.dylib' (no such file), '/usr/local/lib/libSystem.B.dylib' (no such file)
```

**Comment by smaret** [comment]
It bisects to #b40eaa1675

**Comment by smaret** [comment]
To avoid duplicating the work: I'm running a bisect, but its takes a while: many packages needs to be rebuild at each step.

**Comment by smaret** [comment]
It bisects to 76488857ab523252a1e785e1d97e95534865dd4c.

**Comment by K900** [comment]
Time to bisect Python itself then?

**Comment by smaret** [comment]
[Python 3.10.2](https://www.python.org/downloads/release/python-3102/) was been released a couple of days ago. I'm trying to compile it to see if it fixes this issue.

**Comment by smaret** [comment]
Same issue with 3.10.2...

**Comment by veprbl** [comment]
Looking at https://docs.python.org/release/3.10.1/whatsnew/changelog.html#python-3-10-1-final
I get a feeling that the https://bugs.python.org/issue34602 is the culpit
Will attempt bisecting python still

**Comment by layday** [comment]
That doesn't seem relevant and Nix is not using `--with-undefined-behavior-sanitizer` or `--with-pydebug`.

**Comment by veprbl** [comment]
First bad commit is https://github.com/python/cpython/commit/2563dd2d0a1cf793afca328ae9e195b72bd2b391

**Comment by veprbl** [comment]
```
substituteStream(): WARNING: pattern '-Wl,-stack_size,1000000' doesn't match anything in file 'configure'
```
The option was patched away in #18025 since a long time, but for a completely different reason.

**Comment by veprbl** [comment]
Bisected the stack size setting breakage to c595d5ed3dd05e278c12e740d32bcb3650b6f069

**Comment by veprbl** [comment]
There are only 'skip'ped commits left to test.
The first bad commit could be any of:
https://github.com/tpoechtrager/cctools-port/commit/9370fb2beedf80e2a399ed2210ca7dcf7ed77351
https://github.com/tpoechtrager/cctools-port/commit/05dea216c9519492f3d3e4bb8700b7fea812306b
https://github.com/tpoechtrager/cctools-port/commit/1bcd6f8a768588046b9f7c9b6746abfc98a952e6
https://github.com/tpoechtrager/cctools-port/commit/b7f4bf37ce02098e6960e29d13999d046bdf57c5
We cannot bisect more!

**Comment by takeda** [comment]
@veprbl are there any ideas why is this happening?

**Comment by veprbl** [comment]
I don't know why the problem happens, but there is a fix.
This will be fixed once 49a0059a59a950c9261325fad14a2f905f53036d lands on master
